src/aten/sycl/ReduceSumProdKernel.cpp

```diff
@@ -0,0 +1,66 @@
+#include <ATen/Dispatch.h>
+#include <ATen/NumericUtils.h>
+#include <ATen/OpMathType.h>
+#include <ATen/native/DispatchStub.h>
+#include <ATen/native/ReduceAllOps.h>
+#include <ATen/native/ReduceOps.h>
+#include <ATen/native/SharedReduceOps.h>
+#include <ATen/native/TensorCompare.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cuda/ReduceOps.h>
+#include <aten/sycl/NumericLimits.h>
+#include <aten/sycl/Reduce.h>
+
+namespace at {
+namespace native {
+namespace xpu {
+
+template <typename acc_t>
+struct SumFunctor {
+  inline acc_t operator()(acc_t a, acc_t b) const {
+    return a + b;
+  }
+};
+
+template <>
+struct SumFunctor<c10::complex<at::Half>> {
+  using scalar_t = c10::complex<at::Half>;
+  using acc_t = at::opmath_type<scalar_t>;
+  inline acc_t operator()(acc_t a, acc_t b) const {
+    return a + b;
+  }
+};
+
+template <
+    typename scalar_t,
+    typename acc_t = scalar_t,
+    typename out_t = scalar_t>
+struct sum_functor {
```

Comment:
Template function is enough?

Reply:
Either is ok

Reply:
Template class is needed, since need partial specialization.